### PR TITLE
refactor: Add missing includes in pubkey.cpp/pubkey.h

### PR DESCRIPTION
**Pull request by whb07**
#### Problem:
Many symbols in the files were undefined and causing issues when I was working on building independent sections of the codebase. The hidden imports from the "secp256k1" library was a particular pain point.

The other standard and missing includes are following best practices and will help with refactoring, build process and others. 

#### Changes:
Clean up and declared imports/include for `pubkey.cpp` and `pubkey.h`

### CODE REVIEW

**Comment by whb07** [comment]
Would like feedback on the comments next to the imports.

sometimes they can be helpful, but assuming one imports multiple things I’m not sure what styling to take or just omit.

File: src/pubkey.h [hunk 1]
```diff
@@ -7,13 +7,13 @@
...
-#include <uint256.h>
+#include <hash.h>       // for Hash, Hash160
+#include <serialize.h>  // for ReadCompactSize, WriteCompactSize
+#include <span.h>       // for Span, MakeSpan
+#include <uint256.h>    // for u256
 
-#include <stdexcept>
-#include <vector>
+#include <cstring>      // for memcmp, memcpy, size_t
+#include <vector>       // for vector
```
Comment (fanquake) [inline src/pubkey.h:16]: > Would like feedback on the comments next to the imports.

That is not something we want to do. What is the value-add of putting `// for vector` next to `#include <vector>`? In the other cases, these comments would be fairly unmaintainable, frequently become out of date (if not managed in some automated way), would pollute diffs etc.

**Comment by jnewbery** [comment]
I'd much rather we went the other way and removed all of the `// for` comments from include lines, for the reasons of maintainability explained here: https://github.com/bitcoin/bitcoin/pull/21745#discussion_r618111630.

File: src/pubkey.h (hunk 1 above)
Comment (whb07) [reply src/pubkey.h]: I agree. Changes have been implemented by removing comments from includes.

**Comment by whb07** [comment]
> I'd much rather we went the other way and removed all of the `// for` comments from include lines, for the reasons of maintainability explained here: [#21745 (comment)](https://github.com/bitcoin/bitcoin/pull/21745#discussion_r618111630).

Agreed all around then. I've updated the code and removed the "for" comments next to include. Better to include those comments  in the PR itself.

Should be good to go now.

File: src/pubkey.cpp [hunk 2]
```diff
@@ -5,9 +5,17 @@
...
 #include <secp256k1.h>
+#include <secp256k1_extrakeys.h>
 #include <secp256k1_recovery.h>
 #include <secp256k1_schnorrsig.h>
+#include <span.h>
+#include <uint256.h>
+
+#include <algorithm>
+
+#include <assert.h>
```
Comment (jnewbery) [inline src/pubkey.cpp:18]:
```suggestion
#include <cassert>
#include <algorithm>
```

(group all standard library imports together and prefer the c++ versions of the c headers).

**Review by jnewbery — COMMENTED** [review]
Looks better. All commits can be squashed.

File: src/pubkey.cpp (hunk 2 above)
Comment (whb07) [reply src/pubkey.cpp]: Done. Yep that was a question I was slightly mulling over. I'll take a look and see if i see the styling docs (if any).

**Comment by whb07** [comment]
> Looks better. All commits can be squashed.

Commits got squashed. let me know if you need anything else

**Comment by jnewbery** [comment]
utACK 71c824ed6c

Change looks good. Did you use a tool to find the missing includes?

**Comment by whb07** [comment]
> utACK [71c824e](https://github.com/bitcoin/bitcoin/commit/71c824ed6cf70b39ca09e8b3962f452f69523af0)
> 
> Change looks good. Did you use a tool to find the missing includes?

I am working on currently writing up a CMake build for the repo and did run into several issues especially with the secp256k1 libraries due to some different includes within that lib/module.

I generally just crank up all the warnings and flags for the compilers I use and other static analyzers, which also bring in "include what you use" tool.

**Pull request merged**